db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/PredicateList.java

```diff
@@ -564,6 +613,53 @@ public static Integer isIndexUseful(Predicate pred,
         return indexPosition;
     }
 
+    private static boolean matchIndexExpression(RelationalOperator relOp, InListOperatorNode inNode,
+                                                boolean isIn, boolean isProbe,
+                                                ValueNode indexExprAst, Optimizable optTable) {
+        boolean match = false;
+        int tableNumber = optTable.getTableNumber();  // OK to be -1, will be checked when use
+        if (isIn) {
+            if (inNode.getLeftOperand().equals(indexExprAst)) {
+                match = true;
+                if (isProbe) {
+                    assert relOp instanceof BinaryOperatorNode;
+                    BinaryOperatorNode binOp = (BinaryOperatorNode) relOp;
+                    binOp.setMatchIndexExpr(tableNumber, true);
+                }
+            }
+        } else {
+            if (relOp instanceof BinaryOperatorNode) {
+                BinaryOperatorNode binOp = (BinaryOperatorNode) relOp;
+                if (binOp.getLeftOperand().equals(indexExprAst)) {
```

Comment:
The equals method will only match 2 different JavaToSQLValueNodes if they are the exact same object:
>     protected boolean isEquivalent(ValueNode o)
>     {
>         // anything in the java domain is not equiavlent.
>         return false;//this == o;
>     }

Therefore, if you have something like:
> create index t1_idx on t1(ln(a));
> explain select * from t1 --splice-properties index=t1_idx
> where ln(a) > 1;

It cannot take advantage of the index.  It may be better to write a new method for comparing expressions, that defaults to using the equals method for all nodes, but then for JavaToSQLValueNode we could override the implementation and do the proper expression comparison.  Note that non-deterministic functions should not ever be considered the same expression.

Reply:
Absolutely. I just found this yesterday when implementing `hashCode()` and it made my eyes open. There are also some other classes compare using `==`. I'm working on a fix.